### DISCUSSION TITLE
I meet a error that "default"  is not exported.......

### DISCUSSION THREAD

**Discussion by dddssw** [rollup/rollup#5777] [Q&A]
sorry, I just started using rollup and I encountered an error and I don’t know how to solve it

Error [RollupError]: ../../node_modules/.pnpm/element-plus@2.9.1_vue@3.5.13_typescript@5.7.2_/node_modules/element-plus/es/components/time-picker/src/utils.mjs (1:7): "default" 
is not exported by "../../node_modules/.pnpm/dayjs@1.11.13/node_modules/dayjs/dayjs.min.js", imported by "../../node_modules/.pnpm/element-plus@2.9.1_vue@3.5.13_typescript@5.7.2_/node_modules/element-plus/es/components/time-picker/src/utils.mjs".

**Comment by TrickyPi** [comment]
I assume that you are using the `@rollup/plugin-node-resolve` to resolve third-party modules in `node_modules`. Since the `package.json` only includes the `main` field, and its value is `dayjs.min.js`, the `dayjs` import will be resolved to `dayjs/dayjs.min.js`. Unfortunately, `dayjs/dayjs.min.js` is not an ES module. One workaround is to add a custom plugin to resolve `dayjs` to `dayjs/esm/index.js`. Otherwise, you'll need to wait for the `dayjs` maintainer to add `"exports": "esm/index.js"` to the `package.json`.

**Comment by dddssw** [comment]
Thanks, I understand why this problem occurs now .

**Reply by TrickyPi** [reply]
> Thanks, I understand why this problem occurs now .

Oh, there is something I overlooked. The `dayjs.min.js` file can be run as a CJS module, so you can use the `@rollup/plugin-commonjs` plugin to convert CJS to ES Modules to solve this problem. I guess there is also a tip message in your console.
![image](https://github.com/user-attachments/assets/e55c9cfd-8ce1-4404-81b7-f53d92a1c073)
https://rollupjs.org/troubleshooting/#error-name-is-not-exported-by-module

**Reply by dddssw** [reply]
I asked gpt and it told me to use @rollup/plugin-commonjs, but the configuration it provided was not suitable for me 

```
    commonjs({
        include: "node_modules/**",
      }), 
```
Your answer reminded me that this is possible, but configuration is wrong. Just delete the include.

Thank again

**Comment by tt-a1i** [comment]
This issue is caused by the module format of the dayjs package. As @TrickyPi mentioned, dayjs only has `main: "dayjs.min.js"` in its package.json, which is CommonJS/UMD format, but element-plus ESM expects an ESM default export.

**Solution 1: Use `@rollup/plugin-alias` to redirect to ESM version**

dayjs actually ships with an ESM version at `dayjs/esm/index.js`, it just isn't declared in package.json.

```js
import alias from '@rollup/plugin-alias';

export default {
  plugins: [
    alias({
      entries: [
        { find: 'dayjs', replacement: 'dayjs/esm/index.js' }
      ]
    }),
    // ... other plugins
  ]
};
```

**Solution 2: Custom resolve plugin**

```js
const resolveDayjsEsm = {
  name: 'resolve-dayjs-esm',
  resolveId(source) {
    if (source === 'dayjs') {
      return { id: 'dayjs/esm/index.js', external: false };
    }
    return null;
  }
};

export default {
  plugins: [
    resolveDayjsEsm,
    // ... other plugins (put this BEFORE node-resolve)
  ]
};
```

**Note**: If you also use dayjs plugins (e.g., `dayjs/plugin/customParseFormat`), you'll need to map them to their ESM paths as well: `dayjs/esm/plugin/customParseFormat`.

**Comment by SX110903** [comment]
ese error pasa porque dayjs no exporta un default en el formato que espera element-plus. es un problema conocido de compatibilidad entre cjs y esm. en rollup puedes solucionarlo con el plugin commonjs:

```js
import commonjs from "@rollup/plugin-commonjs"

plugins: [
  commonjs({
    include: /node_modules/,
    requireReturnsDefault: "auto"
  })
]
```

la opcion requireReturnsDefault en auto hace que cuando un modulo cjs no tiene default export lo sintetice automaticamente. si eso no funciona prueba con esmExternals: true tambien. es un lio este tema de cjs vs esm con dayjs porque segun la version que uses cambia como exporta
